projects/jupyter-server-ydoc/jupyter_server_ydoc/handlers.py

```diff
@@ -600,3 +602,95 @@ async def _cleanup_undo_manager(self, room_id: str) -> None:
         if room_id in FORK_DOCUMENTS:
             del FORK_DOCUMENTS[room_id]
             self.log.info(f"Fork Document for {room_id} has been removed.")
+
+
+class DocForkHandler(APIHandler):
+    """
+    Jupyter Server handler to:
+    - create a fork of a root document (optionally synchronizing with the root document),
+    - delete a fork of a root document (optionally merging back in the root document).
+    - get fork IDs of a root document.
+    """
+
+    auth_resource = "contents"
+
+    def initialize(
+        self,
+        ywebsocket_server: JupyterWebsocketServer,
+    ) -> None:
+        self._websocket_server = ywebsocket_server
+
+    @web.authenticated
+    @authorized
+    async def get(self, root_roomid):
+        """
+        Returns a dictionary of given root room ID to the root's fork room IDs.
+        """
+        self.write(
+            {
+                root_roomid: [
+                    fork_roomid
+                    for fork_roomid, _root_roomid in FORK_ROOMS.items()
+                    if _root_roomid == root_roomid
+                ]
+            }
+        )
+
+    @web.authenticated
+    @authorized
+    async def put(self, root_roomid):
+        """
+        Creates a fork of a root document and returns its ID.
+        Optionally keeps the fork in sync with the root.
+        """
+        fork_roomid = uuid4().hex
+        FORK_ROOMS[fork_roomid] = root_roomid
+        root_room = await self._websocket_server.get_room(root_roomid)
+        update = root_room.ydoc.get_update()
+        fork_ydoc = Doc()
+        fork_ydoc.apply_update(update)
+        model = self.get_json_body()
+        if model.get("sync"):
+            root_room.ydoc.observe(lambda event: fork_ydoc.apply_update(event.update))
+        fork_room = YRoom(ydoc=fork_ydoc)
+        self._websocket_server.rooms[fork_roomid] = fork_room
+        await self._websocket_server.start_room(fork_room)
+        self._emit_fork_event(self.current_user.username, root_roomid, fork_roomid, "create")
+        data = json.dumps(
+            {
+                "sessionId": SERVER_SESSION,
+                "roomId": fork_roomid,
+            }
+        )
+        self.set_status(201)
+        return self.finish(data)
+
+    @web.authenticated
+    @authorized
+    async def delete(self, fork_roomid):
+        """
+        Deletes a forked document, and optionally merges it back in the root document.
+        """
+        root_roomid = FORK_ROOMS[fork_roomid]
+        del FORK_ROOMS[fork_roomid]
+        if int(self.get_query_argument("merge")):
```

Comment:
using `?merge=true` might be nicer than `?merge=1` ?

Reply:
Yes, done in https://github.com/jupyterlab/jupyter-collaboration/pull/382/commits/08276639c3d27d592f17fcf4f95c79291e33d242.